[SPARK-180|170|182] command parsing overhaul

theunkn0wn1 opened the pull request:
This PR is a big one, it overhauls the way most commands are parsed.

Overarching idea is to parse the incoming commands using `pyparsing`, instead of raw string operations or regex.
By using `pyparsing` over regex, its far more readable.
By using `pyparsing` over direct string methods, its much more flexible *especially* for the more complicated commands.

This PR also has bug fixes for SPARK-170 and SPARK-182 as they fell into scope

theunkn0wn1 committed:
Began writing parsing rules for commonly used inputs

theunkn0wn1 committed:
parse case numbers into native format
 - deprecates validate

theunkn0wn1 committed:
fixed UUID parsing error

theunkn0wn1 committed:
parsing is fun i said...

theunkn0wn1 committed:
dropped the optional `case` because it evals incorrectly -_-

theunkn0wn1 committed:
began porting all of case management commands to pyparsed

theunkn0wn1 committed:
More work towards porting the rest of case management

theunkn0wn1 committed:
you get a token and YOU get a token. EVERYONE gets a token!

theunkn0wn1 committed:
fix some things

theunkn0wn1 committed:
began work towards redoing the manual signal parser

theunkn0wn1 committed:
[SPARK-180 | SPARK-170] minor reworks to the signal handler...

theunkn0wn1 committed:
[SPARK-180 | SPARK-170] better config management

theunkn0wn1 committed:
[SPARK-180 | SPARK-170] use new config object

theunkn0wn1 committed:
[SPARK-170] slight formatting improvements

theunkn0wn1 committed:
[SPARK-182] `cmdr` should read remainder of text instead of an IRC identifier.

Eearslya commented on src/commands/case_management.py:
```diff
@@ -175,32 +215,37 @@ async def cmd_case_management_clear(ctx: Context):
         await ctx.reply(f"Client for {case.board_index} is now CMDR {case.client}")
 
 
+CODE_RED_PATTERN = suppress_first_word + rescue_identifier.setResultsName("subject")
```
[Style] Should this not join its friends at the top of the file?

theunkn0wn1 replied:
probably. will fix

theunkn0wn1 replied:
Moved.

Eearslya commented on src/packages/ratmama/ratmama_parser.py:
```diff
@@ -126,49 +145,47 @@ async def handle_ratmama_announcement(ctx: Context) -> None:
+    )
+    platform_signal = f"({rescue.platform.value.upper()}_SIGNAL)" if rescue.platform else ""
+    await ctx.reply(
+        f"DRILLSIGNAL - CMDR {rescue.client} - "
```
[Error] Hardcoded Drillsignal?

theunkn0wn1 replied:
:tired_face:

theunkn0wn1 replied:
Resolved hardcoded signal.

Eearslya commented on src/packages/ratmama/ratmama_parser.py:
```diff
@@ -232,21 +243,34 @@ async def handle_ratsignal(ctx: Context) -> None:
         else:
             system = part
 
+    if system and "signal" in system.lower():
```
[Question] What case is this meant to catch?

theunkn0wn1 replied:
I honestly don't recall what case this catches,  but it was a buggy edge case.

Eearslya commented on tests/unit/test_ratmama.py:
```diff
@@ -152,8 +152,8 @@ async def test_announcer_reconnect_with_changes(bot_fx, async_callable_fx, monke
-    assert async_callable_fx.was_called_with("System changed! Platform changed!"
-                                             " O2 Status changed, it is now CODE RED!")
+    assert async_callable_fx.was_called_with('Case #0 system, platform, O2 Status changed, '
+                                             'it is now CODE RED! O2 Status changed, it is now CODE RED!')
```
[Question] Is it intentional that O2 status is announced twice?

theunkn0wn1 replied:
no, thats a bug :P

theunkn0wn1 replied:
Resolved by fixing the erroneous underlying logic.

theunkn0wn1 committed:
Fix duplication bug in CR handling
 - fixes some test errors caused by fixed hardcoded signals
 - resolves a bunch of review comments too

theunkn0wn1 committed:
These tests are unreliable and fail health checks frequently.

theunkn0wn1 commented:
I removed two of the hypothesis tests that frequently failed health checks.
It was a good idea to have them, but more effort needs to be made in stabilizing them.

theunkn0wn1 committed:
black -l 102 src/packages/ratmama/ratmama_parser.py src/packages/parsing_rules/__init__.py src/commands/case_management.py

theunkn0wn1 committed:
pycodestyle: chill.

theunkn0wn1 committed:
Actually test parsing rules directly. WIP

theunkn0wn1 committed:
Expand tests, found a bug. WIP.

theunkn0wn1 committed:
Resolved one of the two distinct failures with SUB_CMD_PATTERN
 - reworked some of the strategies to be callables
(i think this is how its supposed to be...)

theunkn0wn1 committed:
Capture the entire word, or it parses incorrectly.

theunkn0wn1 committed:
Fixed sub cmd tests
 - looks like the parser is right here...

theunkn0wn1 committed:
Further expansion of pattern tests
 - after sub its evident the content needs to be checked, not just that the parse succeeded

theunkn0wn1 committed:
converted more of the hypothesis strategies to functions
 - pretty sure this is how the lib was supposed to be used.

theunkn0wn1 committed:
Moar fuzzy tests!
